cpp/src/arrow/compute/light_array.h

```diff
@@ -183,6 +183,36 @@ class ARROW_EXPORT KeyColumnArray {
   // Starting bit offset within the first byte (between 0 and 7)
   // to be used when accessing buffers that store bit vectors.
   int bit_offset_[kMaxBuffers - 1];
+
+  bool is_dictionary_type() const {
```

Comment:
Does the `KeyColumnArray` even support storing dictionary encoded data?  I didn't think it did but I could be remembering incorrectly.

Reply:
Maybe, you are right.

My understanding is that the code is added, but probably unsupported or not being tested properly. 

So, I took a hint from what you said, I removed the code for the dictionary type within `KeyColumnArray` and ran the entire test and found that none of the tests covered that code :O 
(If that's okay, can we add a **new issue** for `KeyColumns` to have complex values like dictionary types?)

Looking at the data I was randomly generating in the `hash_join_node_test.cc` file, I was only generating data for the following types.

In the code below, I have added support for `LargeString` type.

https://github.com/apache/arrow/blob/e9aac8a4ce4b8833667efe52957fb0881a488c4b/cpp/src/arrow/acero/hash_join_node_test.cc#L284-L291

Reply:
>  (If that's okay, can we add a new issue for KeyColumns to have complex values like dictionary types?)

Is this request maybe covered by https://github.com/apache/arrow/issues/31622 , https://github.com/apache/arrow/issues/32494 , or https://github.com/apache/arrow/issues/32504 ?

Adding dictionary support for keys will be trickier than supporting large strings unfortunately.  I can give an overview of some of the steps but I don't know that anyone has been spending a lot of time on the hash join code lately.